src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,173 @@
+
+type AuthSetters =
+  | { type: "passcode"; value: string | null }
+  //This is to set up the future use of the obscure pass beings set by the user. if the `value` is undefined, the default obscure pass is being used
+  | { type: "obscure"; value?: string | null };
+
+type AuthValuesSet = {
+  passcodeSet: boolean;
+  obscureSet: boolean;
+};
+
+type SecurityContextType = {
+  authValuesSet: AuthValuesSet;
+  setAuthValues: (val: AuthSetters) => void;
+  authenticate: (val: string | null, validateOnly?: boolean) => boolean;
+  authState: AuthState;
+};
+
+const DefaultState: SecurityContextType = {
+  authValuesSet: { passcodeSet: false, obscureSet: false },
+  setAuthValues: () => {},
+  authenticate: () => false,
+  authState: "unauthenticated",
+};
+
+export const SecurityContext = React.createContext<SecurityContextType>(
+  DefaultState
+);
+
+const usePersistedPasscodeState = createPersistedState(PASSWORD_KEY);
+const usePersistedObscureState = createPersistedState(OBSCURE_KEY);
+
+export const SecurityProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [passcode, status, setPasscode] = usePersistedPasscodeState<
+    string | null
+  >(null);
+  if (status === "loading") return null;
+  return (
+    <SecurityProviderInner passcode={passcode} setPasscode={setPasscode}>
+      {children}
+    </SecurityProviderInner>
+  );
+};
+
+const SecurityProviderInner = ({
+  children,
+  passcode,
+  setPasscode,
+}: {
+  children: React.ReactNode;
+  passcode: string | null;
+  setPasscode: React.Dispatch<React.SetStateAction<null | string>>;
+}) => {
+  const [authState, setAuthState] = React.useState<AuthState>(
+    passcode === null ? "authenticated" : "unauthenticated"
+  );
+
+  const [obscureCode, , setObscureCode] = usePersistedObscureState<
+    string | null
+  >(null);
+
+  const setPasscodeWithValidation = React.useCallback(
+    (passcodeValue: string | null) => {
+      // If user is able to set their own obscure passcode get rid of this if statement.
+      if (passcodeValue === OBSCURE_PASSCODE) {
+        throw new Error("passcode is reserved");
+      }
+
+      if (passcode !== null && passcodeValue === obscureCode) {
+        throw new Error("passcode is already being used obscure code");
+      }
+
+      if (!validPasscode(passcodeValue)) {
+        throw new Error("passcode not valid");
+      }
+
+      if (passcodeValue === null) {
+        setAuthState("authenticated");
+        setObscureCode(null);
+      }
+
+      setPasscode(passcodeValue);
+    },
+    [obscureCode]
+  );
+
+  const setObscureCodeWithValidation = React.useCallback(
+    (newObscureVal: string | null | undefined) => {
+      if (passcode === null && newObscureVal !== null) {
+        throw new Error("Cannot set obscure mode if passcode not set");
+      }
+
+      if (newObscureVal === undefined) {
+        setObscureCode(OBSCURE_PASSCODE);
+        return;
+      }
```

Comment:
```suggestion
    (newObscureVal: string | null = OBSCURE_PASSCODE) => {
      if (passcode === null && newObscureVal !== null) {
        throw new Error("Cannot set obscure mode if passcode not set");
      }

```

Non-blocking, just wanted to suggest this pattern as an option in the future.